apps/epic-web/src/hooks/use-golden-ticket.tsx

```diff
@@ -0,0 +1,66 @@
+import * as React from 'react'
+import {useRouter} from 'next/router'
+import {trpc} from 'trpc/trpc.client'
+import {useCoupon} from '@skillrecordings/skill-lesson/path-to-purchase/use-coupon'
+
+export type GoldenTicketContextType = {
+  redeemableCoupon?: boolean
+  RedeemDialogForCoupon: () => React.JSX.Element | null
+  validCoupon: boolean
+}
+
+const defaultGoldenTicketContext: GoldenTicketContextType = {
+  redeemableCoupon: false,
+  RedeemDialogForCoupon: () => null,
+  validCoupon: false,
+}
+
+export const GoldenTicketContext = React.createContext(
+  defaultGoldenTicketContext,
+)
+
+export const GoldenTicketProvider = ({
+  children,
+}: {
+  children: React.ReactNode
+}) => {
+  const router = useRouter()
+
+  const {data: couponData} = trpc.coupons.getForCodeOrCoupon.useQuery({
+    code: router.query.code ? (router.query.code as string) : undefined,
+    coupon: router.query.coupon ? (router.query.coupon as string) : undefined,
+  })
+
+  const {data: product} = trpc.products.getProductById.useQuery({
+    productId: couponData?.restrictedToProductId as string,
+  })
```

Comment:
loads the coupon and then the product it's restricted to

Reply:
I was starting to leave a comment about potentially weird behavior with a golden ticket not having a `restrictedToProductId`, but I see you handled that by returning `null` from `getProductById`. It still leaves me with the question of whether it is valid for a golden ticket to be unrestricted to a product and if the provider should have some logic for short-circuiting in that case.